scripts/src/main/java/com/mathewsachin/fategrandautomata/scripts/modules/ServantTracker.kt

```diff
@@ -181,53 +180,51 @@ class ServantTracker @Inject constructor(
         }
     }
 
-    fun faceCardsGroupedByServant(): Map<TeamSlot, List<CommandCard.Face>> {
+    fun faceCardsGroupedByServant(): Map<TeamSlot, Collection<CommandCard.Face>> {
         if (prefs.skipServantFaceCardCheck) {
             return emptyMap()
         }
 
         val cardsRemaining = CommandCard.Face.list.toMutableSet()
-        val result = mutableMapOf<TeamSlot, List<CommandCard.Face>>()
+        val result = mutableMapOf<TeamSlot, MutableSet<CommandCard.Face>>()
 
         supportSlot?.let { supportSlot ->
             if (supportSlot in deployed.values) {
                 val matched = cardsRemaining.filter { card ->
                     images[Images.Support] in locations.attack.supportCheckRegion(card)
-                }
-
-                messages.log(
-                    ScriptLog.CardsBelongToServant(
-                        cards = matched,
-                        servant = supportSlot,
-                        isSupport = true
-                    )
-                )
+                }.toMutableSet()
 
                 cardsRemaining -= matched
 
                 result[supportSlot] = matched
             }
         }
 
-        for (teamSlot in deployed.values) {
-            if (supportSlot != teamSlot) {
-                val img = faceCardImages[teamSlot] ?: continue
-
-                val matched = cardsRemaining.filter { card ->
-                    img in locations.attack.servantMatchRegion(card)
+        val ownedServants = faceCardImages
+            .filterKeys { it != supportSlot && it in deployed.values }
+        for (card in cardsRemaining) {
+            // find the best matching Servant which isn't the support
+            val matchedServants = ownedServants.mapValues { (_, image) ->
+                locations.attack.servantMatchRegion(card)
+                    .find(image, 0.5)?.score ?: 0.0
+            }.filterValues { it > 0.0 }
+
+            matchedServants
+                .maxByOrNull { it.value }
+                ?.let { (teamSlot, _) ->
+                    // add the card to the matched Servant
+                    result.getOrPut(teamSlot, { mutableSetOf() }) += card
```

Comment:
You can use the groupBy operator instead of the
 loop.
Also, see if you can avoid MutableSet and just use Set.